File: day05.py
import sys

def loadInput():
    filename = sys.argv[1] if len(sys.argv) >= 2 else "input05.txt"
    data = open(filename).read()
    lines = [block.split("\n") for block in data.split("\n\n")]
    return parseStack(lines[0]), parseStack(lines[0]), parseCommands(lines[1])


def parseStack(lines):
    numStacks = len(lines[-1].split())
    stacks = [[] for _ in range(numStacks)]
    for line in reversed(lines[:-1]):
        for i, stack in enumerate(stacks):
            if len(line) > i * 4 and line [i * 4] == '[':
                stack.append(line[i * 4 + 1])
    return stacks


def parseCommands(lines):
    commands = []
    for line in lines:
        if not line:
            break
        tokens = line.split()
        commands.append(tuple(int(tokens[i]) for i in [1, 3, 5]))
    return commands
    
if __name__ == "__main__":
    stacks9000, stacks9001, commands = loadInput()
    for n, origin, destination in commands:
        stacks9000[destination - 1].extend(reversed(stacks9000[origin - 1][-n:]))
        stacks9001[destination - 1].extend(stacks9001[origin - 1][-n:])
        stacks9000[origin - 1] = stacks9000[origin - 1][:-n]
        stacks9001[origin - 1] = stacks9001[origin - 1][:-n]
    print("Part 1:", "".join([stack[-1] if stack else ' ' for stack in stacks9000]))
    print("Part 2:", "".join([stack[-1] if stack else ' ' for stack in stacks9001]))

File: day05.swift

import Foundation

struct Stack<T> {
  private var data = [T]()

  public mutating func push(_ val: T) {
    data.append(val)
  }

  public mutating func pop() -> T {
    let result = top()!
    data.removeLast()
    return result
  }

  public func top() -> T? {
    return data.last
  }
}

func buildStacks(_ stackInput: [Substring.SubSequence]) -> [Stack<Character>] {
  // Process input in reversed order (starting at bottom of stacks)
  let stackData = stackInput.reversed().dropFirst(1)
  // Each of the n stacks in the input is 3 characters wide, plus (n-1) spaces in between
  let n = (stackData.first!.count + 1) / 4
  // Parse line by line, build all stacks simultaneously
  var stacks = Array(repeating: Stack<Character>(), count: n)
  for line in stackData {
    for pos in 0..<n {
      let crate = line[line.index(line.startIndex, offsetBy: 4 * pos + 1)]
      if crate != " " { stacks[pos].push(crate) }
    }
  }
  return stacks
}

typealias Step = (Int, Int, Int)

func buildSteps(_ stepInput: [Substring.SubSequence]) -> [Step] {
  // Each step is a line like "move 1 from 2 to 1".
  // Pick numeric items from the line, adjust offsets to start from 0 instead of 1.
  return stepInput.map { $0.split(separator: " ").compactMap { Int($0) } }.map { numbers in
    (numbers[0], numbers[1] - 1, numbers[2] - 1)
  }
}

func combineStackTops(_ stacks: [Stack<Character>]) -> String {
  return String(stacks.map { $0.top() ?? Character("_") })
}

let filename = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "example05.txt"
let blocks = try! String(contentsOfFile: filename)
  .split(separator: "\n\n").map { $0.split(separator: "\n") }
let steps = buildSteps(blocks[1])

// Part 1: move one block at a time (this reverses the order if several blocks are moved at once)
var stacks = buildStacks(blocks[0])
for (n, from, to) in steps {
  for _ in 1...n { stacks[to].push(stacks[from].pop()) }
}
print("Part 1: ", combineStackTops(stacks))

// Part 2: move several blocks at once (to maintain their order, reverse them twice)
stacks = buildStacks(blocks[0])
for (n, from, to) in steps {
  var temp = Stack<Character>()
  for _ in 1...n { temp.push(stacks[from].pop()) }
  for _ in 1...n { stacks[to].push(temp.pop()) }
}
print("Part 2: ", combineStackTops(stacks))
